app/rdf/query-hierarchies.ts

```diff
@@ -0,0 +1,88 @@
+import { DESCRIBE } from "@tpluscode/sparql-builder";
+import clownface from "clownface";
+import { ascending } from "d3";
+import { uniqBy } from "lodash";
+import rdf from "rdf-ext";
+import { DatasetCore, NamedNode, Quad, Stream } from "rdf-js";
+
+import { HierarchyValue } from "@/graphql/resolver-types";
+
+import * as ns from "./namespace";
+import { sparqlClientStream } from "./sparql-client";
+
+const fromStream = (
+  dataset: DatasetCore<Quad, Quad>,
+  stream: Stream<Quad>
+): Promise<DatasetCore<Quad, Quad>> => {
+  return new Promise((resolve) => {
+    stream.on("data", (quad: Quad) => dataset.add(quad));
+    stream.on("end", () => resolve(dataset));
+  });
+};
+
+const hasValueAndLabel = (
+  o: Record<string, any>
+): o is Record<string, string> & { label: string; value: string } => {
+  return o.label && o.value ? true : false;
+};
+
+export const queryHierarchy = async (
+  dimension: string
+): Promise<HierarchyValue[]> => {
+  const query = DESCRIBE`?hierarchy ?level1 ?level2`.WHERE`    
+      SELECT DISTINCT ?hierarchy ?level1 ?level2
+  
+      WHERE {
+    
+        ?cube ${ns.cube.observationConstraint} ?shape.
+        ?shape ?prop ?blankNode.
+        ?blankNode ${ns.sh.path} <${dimension}>.
+        ?blankNode ${ns.cubeMeta.hasHierarchy} ?hierarchy.
+        ?hierarchy ${ns.cubeMeta.hierarchyRoot} ?level1.
+        ?level1 ${ns.schema.hasPart} ?level2 .
+    
+        filter(isiri(?level2))
+    
+        OPTIONAL { ?this ${ns.schema.name} ?order0 } 
+      }
+    `;
+  const stream = await query.execute(sparqlClientStream.query);
+  const dataset = await fromStream(rdf.dataset(), stream);
+  const cf = clownface({ dataset });
```

Comment:
It this way more efficient than the way we did this in the past (just query.execute and the parsing the data)? Or was there an other need to do so?